rpcs3/Emu/Cell/Modules/cellImeJp.cpp

```diff
@@ -1,283 +1,1059 @@
-	cellImeJp.error("cellImeJpOpen()");
+	if (!cursor || cursor > CELL_IMEJP_STRING_MAXLENGTH || cursor > input_string.length())
+		return false;
+
+	input_string.erase(--cursor);
+	cursor_end = cursor;
+
+	if (input_string.empty())
+		input_state = CELL_IMEJP_BEFORE_INPUT;
+
+	return true;
+}
+
+bool ime_jp_manager::deleteWord()
+{
+	if (cursor >= (CELL_IMEJP_STRING_MAXLENGTH - 1) || cursor > (input_string.length() - 1))
+		return false;
+
+	input_string.erase(cursor);
+	cursor_end = cursor;
+
+	if (input_string.empty())
+		input_state = CELL_IMEJP_BEFORE_INPUT;
+
+	return true;
+}
+
+void ime_jp_manager::moveCursor(s8 amount)
+{
+	if (amount > 0)
+	{
+		cursor = std::min(cursor + amount, input_string.length() - 1);
+		cursor_end = std::min(cursor_end + amount, input_string.length() - 1);
+	}
+	else if (amount < 0)
+	{
+		cursor = std::max(0, (s32)cursor + amount);
+		cursor_end = std::max((s32)cursor, (s32)cursor_end + amount);
+	}
+}
+
+void ime_jp_manager::moveCursorEnd(s8 amount)
+{
+	if (amount > 0)
+	{
+		cursor_end = std::max((s32)cursor, (s32)std::min(cursor_end + amount, input_string.length() - 1));
+	}
+	else if (amount < 0)
+	{
+		cursor_end = std::max((s32)cursor, (s32)cursor_end + amount);
+	}
+}
+
+error_code cellImeJpOpen(sys_memory_container_t container_id, vm::ptr<CellImeJpHandle> hImeJpHandle, vm::cptr<CellImeJpAddDic> addDicPath)
+{
+	cellImeJp.todo("cellImeJpOpen(container_id=*0x%x, hImeJpHandle=*0x%x, addDicPath=*0x%x)", container_id, hImeJpHandle, addDicPath);
+
+	if (!container_id || !hImeJpHandle)
+	{
+		return CELL_IMEJP_ERROR_PARAM;
+	}
+
+	const auto manager = g_fxo->get<ime_jp_manager>();
+
+	if (container_id == manager->container_id)
+	{
+		return CELL_IMEJP_ERROR_ALREADY_OPEN;
+	}
+
+	if (addDicPath && addDicPath->path)
+	{
+		cellImeJp.warning("cellImeJpOpen dictionary path = %s", addDicPath->path);
+		manager->dictionary_paths.push_back(std::string(addDicPath->path));
+	}
+
+	manager->container_id = container_id;
+
+	*hImeJpHandle = vm::var<u32>(container_id);
+
 	return CELL_OK;
 }
 
-s32 cellImeJpOpen2()
+error_code cellImeJpOpen2(sys_memory_container_t container_id, vm::ptr<CellImeJpHandle> hImeJpHandle, vm::cptr<CellImeJpAddDic> addDicPath)
 {
-	std::memset(s_ime_string, 0, sizeof(s_ime_string));
-	cellImeJp.error("cellImeJpOpen2()");
+	cellImeJp.todo("cellImeJpOpen2(container_id=*0x%x, hImeJpHandle=*0x%x, addDicPath=*0x%x)", container_id, hImeJpHandle, addDicPath);
+
+	if (!container_id || !hImeJpHandle)
+	{
+		return CELL_IMEJP_ERROR_PARAM;
+	}
+
+	const auto manager = g_fxo->get<ime_jp_manager>();
+
+	if (container_id == manager->container_id)
+	{
+		return CELL_IMEJP_ERROR_ALREADY_OPEN;
+	}
+
+	if (addDicPath && addDicPath->path)
+	{
+		cellImeJp.warning("cellImeJpOpen2 dictionary path = %s", addDicPath->path);
+		manager->dictionary_paths.push_back(std::string(addDicPath->path));
+	}
+
+	manager->container_id = container_id;
+
+	*hImeJpHandle = vm::var<u32>(container_id);
```

Comment:
You got horrible stack corruption right there.

Reply:
how to fix

Reply:
use idm to generate "dummy addresses" and treat them as handles.
games shouldnt sccess yhid memory anyways because we dont know what it should contain.

Reply:
Idm is foreign to me

Reply:
reverse engineer rpcs3 code with it then.

Reply:
be more specific with what you misunderstand with it, i cant help uou much otherwise.

Reply:
isn't it just one or two lines of code? Can't you just make a code suggestion instead of being ominous?

Reply:
in a header: 
```
struct imejp_handle
{
  static const u32 id_base = 0xf0000000;
  static const u32 id_count = 0x1000;
  static const u32 id_step = 0x100;
  const u32 container_id;
};
```
here : `*hImeJpHandle vm::cast(idm::make<imejp_handle>(container_id));`